cargo-miri/src/main.rs

```diff
@@ -0,0 +1,95 @@
+#![feature(let_else)]
+#![allow(clippy::useless_format, clippy::derive_partial_eq_without_eq)]
+
+mod phases;
+mod setup;
+mod util;
+mod version;
+
+use std::{env, iter};
+
+use crate::{phases::*, util::*};
+
+fn main() {
+    // Rustc does not support non-UTF-8 arguments so we make no attempt either.
+    // (We do support non-UTF-8 environment variables though.)
+    let mut args = std::env::args();
+    // Skip binary name.
+    args.next().unwrap();
+
+    // Dispatch to `cargo-miri` phase. Here is a rough idea of "who calls who".
+    //
+    // Initially, we are invoked as `cargo-miri miri run/test`. We first run the setup phase:
+    // - We call `xargo`, and set `RUSTC` back to us, together with `MIRI_CALLED_FROM_XARGO`,
+    //   so that xargo's rustc invocations end up in `phase_rustc` with `RustcPhase::Setup`.
+    //   There we then call the Miri driver with `MIRI_BE_RUSTC` to perform the actual build.
+    //
+    // Then we call `cargo run/test`, exactly forwarding all user flags, plus some configuration so
+    // that we control every binary invoked by cargo:
+    // - We set RUSTC_WRAPPER to ourselves, so for (almost) all rustc invocations, we end up in
+    //   `phase_rustc` with `RustcPhase::Build`. This will in turn either determine that a
+    //   dependency needs to be built (for which it invokes the Miri driver with `MIRI_BE_RUSTC`),
+    //   or determine that this is a binary Miri should run, in which case we generate a JSON file
+    //   with all the information needed to build and run this crate.
+    //   (We don't run it yet since cargo thinks this is a build step, not a run step -- running the
+    //   binary here would lead to a bad user experience.)
+    // - We set RUSTC to the Miri driver and also set `MIRI_BE_RUSTC`, so that gets called by build
+    //   scripts (and cargo uses it for the version query).
+    // - We set `target.*.runner` to `cargo-miri runner`, which ends up calling `phase_runner` for
+    //   `RunnerPhase::Cargo`. This parses the JSON file written in `phase_rustc` and then invokes
+    //   the actual Miri driver for interpretation.
+    // - We set RUSTDOC to ourselves, which ends up in `phase_rustdoc`. There we call regular
+    //   rustdoc with some extra flags, and we set `MIRI_CALLED_FROM_RUSTDOC` to recognize this
+    //   phase in our recursive invocations:
+    //   - We set the `--test-builder` flag of rustdoc to ourselves, which ends up in `phase_rustc`
+    //     with `RustcPhase::Rustdoc`. There we perform a check-build (needed to get the expected
+    //     build failures for `compile_fail` doctests) and then store a JSON file with the
+    //     information needed to run this test.
+    //   - We also set `--runtool` to ourselves, which ends up in `phase_runner` with
+    //     `RunnerPhase::Rustdoc`. There we parse the JSON file written in `phase_rustc` and invoke
+    //     the Miri driver for interpretation.
```

Comment:
@saethlin since you recently dug into cargo-miri a bit, could I ask you to check if this comment is (or would have been) helpful?

@oli-obk please also take a look once you are back. (I'll probably land this before because of conflicts, but we can easily tweak the comment later.)